cpp/arcticdb/version/local_versioned_engine.cpp

```diff
@@ -1008,33 +1015,44 @@ std::vector<std::pair<VersionedItem, FrameAndDescriptor>> LocalVersionedEngine::
     return folly::collect(results_fut).get();
 }
 
-std::vector<std::pair<VersionedItem, FrameAndDescriptor>> LocalVersionedEngine::temp_batch_read_internal_direct(
+std::vector<std::optional<ReadVersionOutput>> LocalVersionedEngine::temp_batch_read_internal_direct(
     const std::vector<StreamId> &stream_ids,
     const std::vector<VersionQuery> &version_queries,
     std::vector<ReadQuery> &read_queries,
     const ReadOptions &read_options) {
     py::gil_scoped_release release_gil;
+    // This read option should always be set when calling batch_read
+    internal::check<ErrorCode::E_ASSERTION_FAILURE>(read_options.batch_throw_on_missing_version_.has_value(),
+                                                    "ReadOptions::batch_throw_on_missing_version_ should always be set here");
     auto versions = batch_get_versions_async(store(), version_map(), stream_ids, version_queries);
-    std::vector<folly::Future<std::pair<VersionedItem, FrameAndDescriptor>>> output;
+    std::vector<folly::Future<std::optional<ReadVersionOutput>>> output;
     for (auto &&version : folly::enumerate(versions)) {
         auto read_query = read_queries.empty() ? ReadQuery{} : read_queries[version.index];
-        output.emplace_back(std::move(*version).thenValue([store = store()](auto maybe_index_key) -> ReadKeyOutput {
-            util::check(static_cast<bool>(maybe_index_key), "Version not found for stream");
-            return store->read_sync(maybe_index_key.value());
-        }).thenValue([store = store(), read_query = std::move(read_query), read_options](auto &&key_segment_pair) {
-            auto [index_key, index_segment] = std::move(key_segment_pair);
-            return async_read_direct(store,
-                                     std::move(index_key),
-                                     std::move(index_segment),
-                                     read_query,
-                                     std::make_shared<BufferHolder>(),
-                                     read_options);
+        output.emplace_back(std::move(*version).thenValue([store = store(), read_options](auto maybe_index_key) -> std::optional<ReadKeyOutput> {
```

Comment:
Minor: avoid copying the entire `read_options` for one boolean.

You can also choose between two different lambdas based on the option instead of passing the option value in for more efficiency.

Reply:
`read_options` is needed by `async_read_direct`, not just for one bool